# Machine Learning on Spark – MLLib

As organizations create more diverse and more user-focused data products and services, there is a growing need for machine learning, which can be used to develop personalizations, recommendations, and predictive insights. The Apache Spark machine learning library (MLlib) allows data scientists to focus on their data problems and models instead of solving the complexities surrounding distributed data (such as infrastructure, configurations, and so on).

From the inception of the Apache Spark project, MLlib was considered foundational for Spark’s success. The key benefit of MLlib is that it allows data scientists to focus on their data problems and models instead of solving the complexities surrounding distributed data (such as infrastructure, configurations, and so on). The data engineers can focus on distributed systems engineering using Spark’s easy-to-use APIs, while the data scientists can leverage the scale and speed of Spark core. Just as important, Spark MLlib is a general-purpose library, providing algorithms for most use cases while at the same time allowing the community to build upon and extend it for specialized use cases.

This week we will look at the basics of some of the spark MLlib packages, next week we build these into machine learning pipelines.

We will start with a databricks database that compares city population to median sale prices of homes.

In [0]:
import pyspark
from pyspark.sql.functions import col
# Use the Spark CSV datasource with options specifying:
# - First line of file is a header
# - Automatically infer the schema of the data
data = spark.read.format("csv").option("header", "true").option("inferSchema", "true").load("/databricks-datasets/samples/population-vs-price/data_geo.csv")
data.cache() # Cache data for faster reuse
data.count()

Out[1]: 294

Take a look at the data schema.

In [0]:
data.printSchema()
data.take(10)

root
 |-- 2014 rank: integer (nullable = true)
 |-- City: string (nullable = true)
 |-- State: string (nullable = true)
 |-- State Code: string (nullable = true)
 |-- 2014 Population estimate: integer (nullable = true)
 |-- 2015 median sales price: double (nullable = true)

Out[2]: [Row(2014 rank=101, City='Birmingham', State='Alabama', State Code='AL', 2014 Population estimate=212247, 2015 median sales price=162.9),
 Row(2014 rank=125, City='Huntsville', State='Alabama', State Code='AL', 2014 Population estimate=188226, 2015 median sales price=157.7),
 Row(2014 rank=122, City='Mobile', State='Alabama', State Code='AL', 2014 Population estimate=194675, 2015 median sales price=122.5),
 Row(2014 rank=114, City='Montgomery', State='Alabama', State Code='AL', 2014 Population estimate=200481, 2015 median sales price=129.0),
 Row(2014 rank=64, City='Anchorage[19]', State='Alaska', State Code='AK', 2014 Population estimate=301010, 2015 median sales price=None),
 Row(2014 rank=78, City='Chandl

Here we clean the data by dropping any rows with missing values.

In [0]:
data = data.dropna() # drop rows with missing values
data.count()

Out[3]: 109

Next we select the columns we will use to predict house prices (features) and the correct output (labels). Below we use the databricks display function which can be useful for examining our dataframes.

In [0]:
dataFeat = data.select(["2014 Population estimate","2015 median sales price"])
display(dataFeat)

2014 Population estimate 2015 median sales price 212247 162.9 188226 157.7 194675 122.5 200481 129.0 1537058 206.1 527972 178.1 197706 131.8 346997 685.7 3928864 434.7 319504 281.0 485199 275.8 1381069 510.3 852469 748.3 1015785 900.0 105112 442.2 445830 220.1 663862 338.1 147612 342.7 124705 202.6 130282 205.1 658893 367.8 169854 208.0 128460 169.9 853382 182.5 102346 136.0 430332 269.1 105838 145.0 174110 172.0 188107 170.5 358699 156.0 456002 158.0 350399 699.3 216282 178.7 2722389 192.5 115828 105.2 149123 78.6 116809 115.2 258522 106.3 848788 139.2 129195 156.9 102448 109.4 127215 105.3 388413 126.0 310797 141.2 612780 141.1 228895 177.2 198242 151.6 622793 223.1 655884 374.6 153991 183.9 183016 226.9 114620 114.6 407207 209.4 171155 154.1 470800 153.0 165378 116.5 272996 150.2 446599 148.1 613599 205.9 236995 268.7 110448 237.4 280579 350.8 557169 172.2 258703 117.3 8491079 388.6 209983 118.9 144263 118.3 809958 197.2 251893 196.9 203948 145.0 282586 138.3 439896 218.9 113657 197.5 239269 138.6 115863 180.0 197859 95.4 298165 135.0 389521 105.9 835957 149.9 141003 107.2 281031 83.8 399682 145.1 160561 211.3 619360 289.4 161637 193.9 119104 162.7 1560297 204.9 179154 233.3 168586 159.5 173778 143.6 184281 149.2 656861 132.1 644014 186.4 120958 145.7 197254 144.4 912791 249.1 117585 133.3 320434 172.9 1281047 192.5 679036 138.4 2239558 200.3 1436697 184.7 190884 243.3 217853 212.5 450980 190.0 668342 352.4 212052 180.2 245691 226.2 599642 195.4

Finally we need to convert the data to the format expected by spark, a vector of features as a single column and the label as an additional column.

We define a function that does this (which we can use later). As you can see it takes our DataFrame and converts it to an RDD. This then allows us to apply the map transformation to each entry of the RDD. Here we take all the columns except the last (in our case only one) and convert these to a dense vector, we keep the last column seperate for the label.

Vectors (dense or sparse) are a way of holding data that is used a lot in machine learning. In the features column below we only have one data item (the size of the population) but as you can see the data structure is as follows: [1,1,[],[188226]] this format makes more sense when we use sparse vectors (it easier to see the data is one of the advantages, but mostly its is more space efficient) for more information this video has a short explanation of the vector format: https://www.youtube.com/watch?v=oGwEv82ifrE

Here is the same vector in three formats:
```python
# Use a NumPy array as a dense vector.
dv1 = np.array([1.0, 0.0, 3.0])
# Use a Python list as a dense vector.
dv2 = [1.0, 0.0, 3.0]
# Create a SparseVector.
sv1 = Vectors.sparse(3, [0, 2], [1.0, 3.0]) 
```
You will find out that most of the machine learning algorithms in Spark are based on the features and label format. That is to say, you can play with all of the machine learning algorithms in Spark aftre you prepare your data in this format with features and label.

The function defined below will take a row of our dataframe and convert all of the features into a dense vector. The final dataformat is a dataframe with two columns, a label column and a features column.

https://spark.apache.org/docs/latest/ml-guide.html

In [0]:
from pyspark.sql import Row
from pyspark.ml.linalg import Vectors

In [0]:
a = dataFeat.rdd.map(lambda x: Vectors.dense(x))
a.collect()

Out[6]: [DenseVector([212247.0, 162.9]),
 DenseVector([188226.0, 157.7]),
 DenseVector([194675.0, 122.5]),
 DenseVector([200481.0, 129.0]),
 DenseVector([1537058.0, 206.1]),
 DenseVector([527972.0, 178.1]),
 DenseVector([197706.0, 131.8]),
 DenseVector([346997.0, 685.7]),
 DenseVector([3928864.0, 434.7]),
 DenseVector([319504.0, 281.0]),
 DenseVector([485199.0, 275.8]),
 DenseVector([1381069.0, 510.3]),
 DenseVector([852469.0, 748.3]),
 DenseVector([1015785.0, 900.0]),
 DenseVector([105112.0, 442.2]),
 DenseVector([445830.0, 220.1]),
 DenseVector([663862.0, 338.1]),
 DenseVector([147612.0, 342.7]),
 DenseVector([124705.0, 202.6]),
 DenseVector([130282.0, 205.1]),
 DenseVector([658893.0, 367.8]),
 DenseVector([169854.0, 208.0]),
 DenseVector([128460.0, 169.9]),
 DenseVector([853382.0, 182.5]),
 DenseVector([102346.0, 136.0]),
 DenseVector([430332.0, 269.1]),
 DenseVector([105838.0, 145.0]),
 DenseVector([174110.0, 172.0]),
 DenseVector([188107.0, 170.5]),
 DenseVector([358699.0, 156.0])

In [0]:
def transData(data):
  # Combine columns to a dense vector (excluding the last column)
  dataFeaturesRDD = data.rdd.map(lambda r: [Vectors.dense(r[:-1]),r[-1]])
  
  # Convert the RDD back to a DataFrame, labelling the columns
  featuresDF =  dataFeaturesRDD.toDF(['features','label'])
  
  return featuresDF

In [0]:
dataLR = transData(dataFeat)
dataLR.show()

+-----------+-----+
|   features|label|
+-----------+-----+
| [212247.0]|162.9|
| [188226.0]|157.7|
| [194675.0]|122.5|
| [200481.0]|129.0|
|[1537058.0]|206.1|
| [527972.0]|178.1|
| [197706.0]|131.8|
| [346997.0]|685.7|
|[3928864.0]|434.7|
| [319504.0]|281.0|
| [485199.0]|275.8|
|[1381069.0]|510.3|
| [852469.0]|748.3|
|[1015785.0]|900.0|
| [105112.0]|442.2|
| [445830.0]|220.1|
| [663862.0]|338.1|
| [147612.0]|342.7|
| [124705.0]|202.6|
| [130282.0]|205.1|
+-----------+-----+
only showing top 20 rows



Finally we can run our linear regression. Here we create two models with different parameters and print the coefficients they provide.

In [0]:
# Import LinearRegression class
from pyspark.ml.regression import LinearRegression

# Define LinearRegression algorithm
lr = LinearRegression()

# Fit 2 models, using different regularization parameters
modelA = lr.fit(dataLR, {lr.regParam:0.0})
modelB = lr.fit(dataLR, {lr.regParam:100.0})

# Print the fitted model parameters
print(">>>> ModelA intercept: %r, coefficient: %r" % (modelA.intercept, modelA.coefficients[0]))
print(">>>> ModelB intercept: %r, coefficient: %r" % (modelB.intercept, modelB.coefficients[0]))

>>>> ModelA intercept: 191.29427575139394, coefficient: 3.779789682338248e-05
>>>> ModelB intercept: 199.85112564667153, coefficient: 2.160349948371716e-05


Sometimes we will want to see the significance of the coefficients we have. We can check the p values of our regressors:

In [0]:
seA = modelA.summary.pValues
seB = modelB.summary.pValues

print(">>>> ModelA p value intercept: %r, p value coefficient: %r" % (seA[1], seA[0]))
print(">>>> ModelB p value intercept: %r, p value coefficient: %r" % (seB[1], seB[0]))

>>>> ModelA p value intercept: 0.0, p value coefficient: 0.005212735655699907
>>>> ModelB p value intercept: 0.0, p value coefficient: 0.0345130599830199


Generating predictions based on the model is easy, here we use the input data and compare model predicted to actual values.

In [0]:
predictionsA = modelA.transform(dataLR)
predictionsB = modelB.transform(dataLR)

In [0]:
display(predictionsA)

features label prediction Map(vectorType -> dense, length -> 1, values -> List(212247.0)) 162.9 199.3167659584664 Map(vectorType -> dense, length -> 1, values -> List(188226.0)) 157.7 198.40882267887193 Map(vectorType -> dense, length -> 1, values -> List(194675.0)) 122.5 198.65258131548592 Map(vectorType -> dense, length -> 1, values -> List(200481.0)) 129.0 198.87203590444247 Map(vectorType -> dense, length -> 1, values -> List(1537058.0)) 206.1 249.39183544694856 Map(vectorType -> dense, length -> 1, values -> List(527972.0)) 178.1 211.25050693302884 Map(vectorType -> dense, length -> 1, values -> List(197706.0)) 131.8 198.7671467407576 Map(vectorType -> dense, length -> 1, values -> List(346997.0)) 685.7 204.4100325554172 Map(vectorType -> dense, length -> 1, values -> List(3928864.0)) 434.7 339.79707185649573 Map(vectorType -> dense, length -> 1, values -> List(319504.0)) 281.0 203.37085497805194 Map(vectorType -> dense, length -> 1, values -> List(485199.0)) 275.8 209.63377749220228 Map(vectorType -> dense, length -> 1, values -> List(1381069.0)) 510.3 243.49577931936597 Map(vectorType -> dense, length -> 1, values -> List(852469.0)) 748.3 223.51581105852597 Map(vectorType -> dense, length -> 1, values -> List(1015785.0)) 900.0 229.6888123761335 Map(vectorType -> dense, length -> 1, values -> List(105112.0)) 442.2 195.26728828229332 Map(vectorType -> dense, length -> 1, values -> List(445830.0)) 220.1 208.14571209216254 Map(vectorType -> dense, length -> 1, values -> List(663862.0)) 338.1 216.3868631323583 Map(vectorType -> dense, length -> 1, values -> List(147612.0)) 342.7 196.87369889728708 Map(vectorType -> dense, length -> 1, values -> List(124705.0)) 202.6 196.00786247475386 Map(vectorType -> dense, length -> 1, values -> List(130282.0)) 205.1 196.21866134533786 Map(vectorType -> dense, length -> 1, values -> List(658893.0)) 367.8 216.1990453830429 Map(vectorType -> dense, length -> 1, values -> List(169854.0)) 208.0 197.71439971843276 Map(vectorType -> dense, length -> 1, values -> List(128460.0)) 169.9 196.14979357732565 Map(vectorType -> dense, length -> 1, values -> List(853382.0)) 182.5 223.55032053832574 Map(vectorType -> dense, length -> 1, values -> List(102346.0)) 136.0 195.16273929967986 Map(vectorType -> dense, length -> 1, values -> List(430332.0)) 269.1 207.55992028719376 Map(vectorType -> dense, length -> 1, values -> List(105838.0)) 145.0 195.2947295553871 Map(vectorType -> dense, length -> 1, values -> List(174110.0)) 172.0 197.87526756731307 Map(vectorType -> dense, length -> 1, values -> List(188107.0)) 170.5 198.40432472914995 Map(vectorType -> dense, length -> 1, values -> List(358699.0)) 156.0 204.8523435440444 Map(vectorType -> dense, length -> 1, values -> List(456002.0)) 158.0 208.53019229865 Map(vectorType -> dense, length -> 1, values -> List(350399.0)) 699.3 204.53862100041033 Map(vectorType -> dense, length -> 1, values -> List(216282.0)) 178.7 199.46928047214874 Map(vectorType -> dense, length -> 1, values -> List(2722389.0)) 192.5 294.19485428650535 Map(vectorType -> dense, length -> 1, values -> List(115828.0)) 105.2 195.67233054465268 Map(vectorType -> dense, length -> 1, values -> List(149123.0)) 78.6 196.9308115193872 Map(vectorType -> dense, length -> 1, values -> List(116809.0)) 115.2 195.70941028143642 Map(vectorType -> dense, length -> 1, values -> List(258522.0)) 106.3 201.06586363396843 Map(vectorType -> dense, length -> 1, values -> List(848788.0)) 139.2 223.37667700031912 Map(vectorType -> dense, length -> 1, values -> List(129195.0)) 156.9 196.17757503149085 Map(vectorType -> dense, length -> 1, values -> List(102448.0)) 109.4 195.16659468515581 Map(vectorType -> dense, length -> 1, values -> List(127215.0)) 105.3 196.10273519578055 Map(vectorType -> dense, length -> 1, values -> List(388413.0)) 126.0 205.9754702502544 Map(vectorType -> dense, length -> 1, values -> List(310797.0)) 141.2 203.04174869041074 Map(vectorType -> dense, length -> 1, values -> List(612780.0)) 141

In [0]:
display(predictionsB)

features label prediction Map(vectorType -> dense, length -> 1, values -> List(212247.0)) 162.9 204.43640360159205 Map(vectorType -> dense, length -> 1, values -> List(188226.0)) 157.7 203.91746594049368 Map(vectorType -> dense, length -> 1, values -> List(194675.0)) 122.5 204.05678690866418 Map(vectorType -> dense, length -> 1, values -> List(200481.0)) 129.0 204.18221682666663 Map(vectorType -> dense, length -> 1, values -> List(1537058.0)) 206.1 233.05695735611488 Map(vectorType -> dense, length -> 1, values -> List(527972.0)) 178.1 211.25716847608865 Map(vectorType -> dense, length -> 1, values -> List(197706.0)) 131.8 204.12226711559933 Map(vectorType -> dense, length -> 1, values -> List(346997.0)) 685.7 207.34747515702293 Map(vectorType -> dense, length -> 1, values -> List(3928864.0)) 434.7 284.7283370422665 Map(vectorType -> dense, length -> 1, values -> List(319504.0)) 281.0 206.7535301457171 Map(vectorType -> dense, length -> 1, values -> List(485199.0)) 275.8 210.3331219926716 Map(vectorType -> dense, length -> 1, values -> List(1381069.0)) 510.3 229.6870490751493 Map(vectorType -> dense, length -> 1, values -> List(852469.0)) 748.3 218.2674392480564 Map(vectorType -> dense, length -> 1, values -> List(1015785.0)) 900.0 221.79563636973916 Map(vectorType -> dense, length -> 1, values -> List(105112.0)) 442.2 202.121912684404 Map(vectorType -> dense, length -> 1, values -> List(445830.0)) 220.1 209.48261382149715 Map(vectorType -> dense, length -> 1, values -> List(663862.0)) 338.1 214.19286802093097 Map(vectorType -> dense, length -> 1, values -> List(147612.0)) 342.7 203.040061412462 Map(vectorType -> dense, length -> 1, values -> List(124705.0)) 202.6 202.5451900497885 Map(vectorType -> dense, length -> 1, values -> List(130282.0)) 205.1 202.66567276640916 Map(vectorType -> dense, length -> 1, values -> List(658893.0)) 367.8 214.08552023199638 Map(vectorType -> dense, length -> 1, values -> List(169854.0)) 208.0 203.52056644797884 Map(vectorType -> dense, length -> 1, values -> List(128460.0)) 169.9 202.62631119034984 Map(vectorType -> dense, length -> 1, values -> List(853382.0)) 182.5 218.28716324308505 Map(vectorType -> dense, length -> 1, values -> List(102346.0)) 136.0 202.06215740483205 Map(vectorType -> dense, length -> 1, values -> List(430332.0)) 269.1 209.14780278649852 Map(vectorType -> dense, length -> 1, values -> List(105838.0)) 145.0 202.1375968250292 Map(vectorType -> dense, length -> 1, values -> List(174110.0)) 172.0 203.61251094178152 Map(vectorType -> dense, length -> 1, values -> List(188107.0)) 170.5 203.9148951240551 Map(vectorType -> dense, length -> 1, values -> List(358699.0)) 156.0 207.6002793079814 Map(vectorType -> dense, length -> 1, values -> List(456002.0)) 158.0 209.70236461824553 Map(vectorType -> dense, length -> 1, values -> List(350399.0)) 699.3 207.42097026226654 Map(vectorType -> dense, length -> 1, values -> List(216282.0)) 178.7 204.52357372200885 Map(vectorType -> dense, length -> 1, values -> List(2722389.0)) 192.5 258.6642550026488 Map(vectorType -> dense, length -> 1, values -> List(115828.0)) 105.2 202.35341578487152 Map(vectorType -> dense, length -> 1, values -> List(149123.0)) 78.6 203.0727043001819 Map(vectorType -> dense, length -> 1, values -> List(116809.0)) 115.2 202.37460881786504 Map(vectorType -> dense, length -> 1, values -> List(258522.0)) 106.3 205.43610554020105 Map(vectorType -> dense, length -> 1, values -> List(848788.0)) 139.2 218.18791676645685 Map(vectorType -> dense, length -> 1, values -> List(129195.0)) 156.9 202.64218976247037 Map(vectorType -> dense, length -> 1, values -> List(102448.0)) 109.4 202.06436096177939 Map(vectorType -> dense, length -> 1, values -> List(127215.0)) 105.3 202.5994148334926 Map(vectorType -> dense, length -> 1, values -> List(388413.0)) 126.0 208.24220569164058 Map(vectorType -> dense, length -> 1, values -> List(310797.0)) 141.2 206.56542847571237 Map(vectorType -> dense, length -> 1, values -> List(612780.0)) 141.1 

The predictions dont look too great, in databricks we can easily plot the residuals. You can drag the chart to make it bigger.

In [0]:
display(modelB, dataLR)
#they should have a random distribution around 0

fitted values residuals 204.43640360159205 -41.53640360159204 203.91746594049368 -46.217465940493696 204.05678690866418 -81.55678690866418 204.18221682666663 -75.18221682666663 233.05695735611488 -26.956957356114884 211.25716847608865 -33.15716847608866 204.12226711559933 -72.32226711559932 207.34747515702293 478.3525248429771 284.7283370422665 149.97166295773349 206.7535301457171 74.2464698542829 210.3331219926716 65.4668780073284 229.6870490751493 280.6129509248507 218.2674392480564 530.0325607519435 221.79563636973916 678.2043636302608 202.121912684404 240.07808731559598 209.48261382149715 10.61738617850284 214.19286802093097 123.90713197906905 203.040061412462 139.659938587538 202.5451900497885 0.054809950211506475 202.66567276640916 2.4343272335908352 214.08552023199638 153.71447976800363 203.52056644797884 4.479433552021163 202.62631119034984 -32.72631119034983 218.28716324308505 -35.78716324308505 202.1375968250292 -57.13759682502919 203.61251094178152 -31.612510941781522 203.9148951240551 -33.41489512405511 207.6002793079814 -51.600279307981396 209.70236461824553 -51.702364618245525 207.42097026226654 491.8790297377334 204.52357372200885 -25.82357372200886 258.6642550026488 -66.16425500264882 202.35341578487152 -97.15341578487151 203.0727043001819 -124.4727043001819 205.43610554020105 -99.13610554020106 218.18791676645685 -78.98791676645686 202.64218976247037 -45.74218976247036 202.06436096177939 -92.66436096177938 202.5994148334926 -97.2994148334926 208.24220569164058 -82.24220569164058 206.56542847571237 -65.36542847571238 213.08931806030373 -71.98931806030373 204.79605866099698 -27.596058660996988 204.1338465913226 -52.5338465913226 214.0205153020499 160.57948469795014 203.17787013566863 -19.27787013566862 202.3273187574952 -87.7273187574952 208.64822186093755 0.751778139062452 203.54867260080715 -49.44867260080716 210.02205320360557 -57.022053203605566 203.4238691842897 -86.92386918428971 205.74879459172837 -55.54879459172838 209.49922691260014 -61.39922691260014 213.1070113263809 -7.2070113263808935 204.97104700681507 63.72895299318492 202.23718895764912 35.16281104235088 205.91261392831342 144.8873860716866 211.88792585051473 -39.687925850514745 205.44001577360763 -88.14001577360763 383.2881464393731 5.311853560626901 204.38749327876093 -85.48749327876092 202.96771129269104 -84.66771129269104 217.34905288150412 -20.149052881504133 205.2928959421235 -8.392895942123488 204.2571161593767 -59.25711615937669 205.95597215177722 -67.6559721517772 209.35441865556078 9.545581344439228 202.3065145874924 -4.806514587492387 205.02017336464107 -66.42017336464107 202.35417190735345 -22.354171907353447 204.12557245102033 -108.72557245102033 206.29253307023407 -71.29253307023407 208.26614236906852 -102.36614236906851 217.9107222645813 -68.01072226458129 202.8972838843741 -95.6972838843741 205.92237871008004 -122.12237871008004 208.48565552732256 -63.38565552732257 203.31980512727665 7.980194872723359 213.2314690869066 76.16853091309338 203.34305049272112 -9.443050492721113 202.42418884918018 -39.724188849180194 233.55900108061695 -28.659001080616946 203.72147899317739 29.578521006822626 203.49317321063347 -43.99317321063347 203.60533857995293 -60.005338579952934 203.83224013503042 -54.63224013503043 214.04162192104548 -81.94162192104548 202.464241737223 -56.76424173722302 204.11250233383268 -59.712502333832674 202.3913731334644 -69.0913731334644 206.77362140023695 -33.87362140023694 227.52622384978895 -35.02622384978895 214.52067952209688 -76.12067952209688 248.23341574342618 -47.93341574342617 230.88880854442954 -46.18880854442955 203.9748880421214 39.32511195787862 204.55751281969776 7.942487180302237 209.5938718438383 -19.593871843838286 214.28965169861803 138.11034830138195 204.43219091919272 -24.232190919192732 205.1589110383255 21.04108896167449 212.80549128408666 -17.40549128408665

Additionaly, Spark includes functionality to help us with our models, for example the regression evaluator.

In [0]:
from pyspark.ml.evaluation import RegressionEvaluator
evaluator = RegressionEvaluator(metricName="rmse")
RMSE_A = evaluator.evaluate(predictionsA)
RMSE_B = evaluator.evaluate(predictionsB)

print("ModelA: Root Mean Squared Error = " + str(RMSE_A))
print("ModelB: Root Mean Squared Error = " + str(RMSE_B))

#Root Mean Square Error (RMSE) is the standard deviation of the residuals (prediction errors). Residuals are a measure of how far from the regression line data points are; RMSE is a measure of how spread out these residuals are. 

ModelA: Root Mean Squared Error = 128.60202684284758
ModelB: Root Mean Squared Error = 129.49630019270606


According to RMSE ModelA performs better.


### Practice exercise

In the below two cells we download a database of car performance (efficiency in miles per gallon MPG) and a number of car characteristics. Use the linear regression estimator to try to predict miles per gallon MPG based on:

 - Just <b> weight </b> as a feature

 - <b> displacement, weight, acceleration, and year </b> as features

Compare the RMSE of your the two aproaches and plot the residuals.

Hint: the `transData` function we defined is very useful, however, it requires the final column to be the data we are trying to predict. You will need to either change this function or re-arrange the dataframe, as current, MPG is the first column.

In [0]:
%sh wget https://raw.githubusercontent.com/plotly/datasets/master/auto-mpg.csv

--2024-05-18 19:30:02--  https://raw.githubusercontent.com/plotly/datasets/master/auto-mpg.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10118 (9.9K) [text/plain]
Saving to: ‘auto-mpg.csv’

     0K .........                                             100% 37.2M=0s

2024-05-18 19:30:02 (37.2 MB/s) - ‘auto-mpg.csv’ saved [10118/10118]



In [0]:
dataMPG = spark.read.format("csv").option("header", "true").option("inferSchema", "true").load("file:/databricks/driver/auto-mpg.csv")
dataMPG.cache()
print(dataMPG.count())
dataMPG.printSchema()

398
root
 |-- mpg: double (nullable = true)
 |-- cylinders: integer (nullable = true)
 |-- displacement: double (nullable = true)
 |-- horsepower: integer (nullable = true)
 |-- weight: integer (nullable = true)
 |-- acceleration: double (nullable = true)
 |-- model-year: integer (nullable = true)



In [0]:
display(dataMPG.select([dataMPG.columns[2], dataMPG.columns[4], dataMPG.columns[5],  dataMPG.columns[6], dataMPG.columns[0]]))

displacement weight acceleration model-year mpg 307.0 3504 12.0 70 18.0 350.0 3693 11.5 70 15.0 318.0 3436 11.0 70 18.0 304.0 3433 12.0 70 16.0 302.0 3449 10.5 70 17.0 429.0 4341 10.0 70 15.0 454.0 4354 9.0 70 14.0 440.0 4312 8.5 70 14.0 455.0 4425 10.0 70 14.0 390.0 3850 8.5 70 15.0 383.0 3563 10.0 70 15.0 340.0 3609 8.0 70 14.0 400.0 3761 9.5 70 15.0 455.0 3086 10.0 70 14.0 113.0 2372 15.0 70 24.0 198.0 2833 15.5 70 22.0 199.0 2774 15.5 70 18.0 200.0 2587 16.0 70 21.0 97.0 2130 14.5 70 27.0 97.0 1835 20.5 70 26.0 110.0 2672 17.5 70 25.0 107.0 2430 14.5 70 24.0 104.0 2375 17.5 70 25.0 121.0 2234 12.5 70 26.0 199.0 2648 15.0 70 21.0 360.0 4615 14.0 70 10.0 307.0 4376 15.0 70 10.0 318.0 4382 13.5 70 11.0 304.0 4732 18.5 70 9.0 97.0 2130 14.5 71 27.0 140.0 2264 15.5 71 28.0 113.0 2228 14.0 71 25.0 98.0 2046 19.0 71 25.0 232.0 2634 13.0 71 19.0 225.0 3439 15.5 71 16.0 250.0 3329 15.5 71 17.0 250.0 3302 15.5 71 19.0 232.0 3288 15.5 71 18.0 350.0 4209 12.0 71 14.0 400.0 4464 11.5 71 14.0 351.0 4154 13.5 71 14.0 318.0 4096 13.0 71 14.0 383.0 4955 11.5 71 12.0 400.0 4746 12.0 71 13.0 400.0 5140 12.0 71 13.0 258.0 2962 13.5 71 18.0 140.0 2408 19.0 71 22.0 250.0 3282 15.0 71 19.0 250.0 3139 14.5 71 18.0 122.0 2220 14.0 71 23.0 116.0 2123 14.0 71 28.0 79.0 2074 19.5 71 30.0 88.0 2065 14.5 71 30.0 71.0 1773 19.0 71 31.0 72.0 1613 18.0 71 35.0 97.0 1834 19.0 71 27.0 91.0 1955 20.5 71 26.0 113.0 2278 15.5 72 24.0 97.5 2126 17.0 72 25.0 97.0 2254 23.5 72 23.0 140.0 2408 19.5 72 20.0 122.0 2226 16.5 72 21.0 350.0 4274 12.0 72 13.0 400.0 4385 12.0 72 14.0 318.0 4135 13.5 72 15.0 351.0 4129 13.0 72 14.0 304.0 3672 11.5 72 17.0 429.0 4633 11.0 72 11.0 350.0 4502 13.5 72 13.0 350.0 4456 13.5 72 12.0 400.0 4422 12.5 72 13.0 70.0 2330 13.5 72 19.0 304.0 3892 12.5 72 15.0 307.0 4098 14.0 72 13.0 302.0 4294 16.0 72 13.0 318.0 4077 14.0 72 14.0 121.0 2933 14.5 72 18.0 121.0 2511 18.0 72 22.0 120.0 2979 19.5 72 21.0 96.0 2189 18.0 72 26.0 122.0 2395 16.0 72 22.0 97.0 2288 17.0 72 28.0 120.0 2506 14.5 72 23.0 98.0 2164 15.0 72 28.0 97.0 2100 16.5 72 27.0 350.0 4100 13.0 73 13.0 304.0 3672 11.5 73 14.0 350.0 3988 13.0 73 13.0 302.0 4042 14.5 73 14.0 318.0 3777 12.5 73 15.0 429.0 4952 11.5 73 12.0 400.0 4464 12.0 73 13.0 351.0 4363 13.0 73 13.0 318.0 4237 14.5 73 14.0 440.0 4735 11.0 73 13.0 455.0 4951 11.0 73 12.0 360.0 3821 11.0 73 13.0 225.0 3121 16.5 73 18.0 250.0 3278 18.0 73 16.0 232.0 2945 16.0 73 18.0 250.0 3021 16.5 73 18.0 198.0 2904 16.0 73 23.0 97.0 1950 21.0 73 26.0 400.0 4997 14.0 73 11.0 400.0 4906 12.5 73 12.0 360.0 4654 13.0 73 13.0 350.0 4499 12.5 73 12.0 232.0 2789 15.0 73 18.0 97.0 2279 19.0 73 20.0 140.0 2401 19.5 73 21.0 108.0 2379 16.5 73 22.0 70.0 2124 13.5 73 18.0 122.0 2310 18.5 73 19.0 155.0 2472 14.0 73 21.0 98.0 2265 15.5 73 26.0 350.0 4082 13.0 73 15.0 400.0 4278 9.5 73 16.0 68.0 1867 19.5 73 29.0 116.0 2158 15.5 73 24.0 114.0 2582 14.0 73 20.0 121.0 2868 15.5 73 19.0 318.0 3399 11.0 73 15.0 121.0 2660 14.0 73 24.0 156.0 2807 13.5 73 20.0 350.0 3664 11.0 73 11.0 198.0 3102 16.5 74 20.0 200.0 2875 17.0 74 21.0 232.0 2901 16.0 74 19.0 250.0 3336 17.0 74 15.0 79.0 1950 19.0 74 31.0 122.0 2451 16.5 74 26.0 71.0 1836 21.0 74 32.0 140.0 2542 17.0 74 25.0 250.0 3781 17.0 74 16.0 258.0 3632 18.0 74 16.0 225.0 3613 16.5 74 18.0 302.0 4141 14.0 74 16.0 350.0 4699 14.5 74 13.0 318.0 4457 13.5 74 14.0 302.0 4638 16.0 74 14.0 304.0 4257 15.5 74 14.0 98.0 2219 16.5 74 29.0 79.0 1963 15.5 74 26.0 97.0 2300 14.5 74 26.0 76.0 1649 16.5 74 31.0 83.0 2003 19.0 74 32.0 90.0 2125 14.5 74 28.0 90.0 2108 15.5 74 24.0 116.0 2246 14.0 74 26.0 120.0 2489 15.0 74 24.0 108.0 2391 15.5 74 26.0 79.0 2000 16.0 74 31.0 225.0 3264 16.0 75 19.0 250.0 3459 16.0 75 18.0 250.0 3432 21.0 75 15.0 250.0 3158 19.5 75 15.0 400.0 4668 11.5 75 16.0 350.0 4440 14.0 75 15.0 318.0 4498 14.5 75 16.0 351.0 4657 13.5 75 14.0 231.0 3907 21.0 75 17.0 250.0 3897 18.5 75 16.0 258.0 3730 19.0 75 15.0 225.0 3785 19.0 75 18.0 231.0 3039 15.0 75 21.0 262.0 3221 13.5 75 20.0 302.0 3

In [0]:
data1 = transData(dataMPG.select([dataMPG.columns[4], dataMPG.columns[0]]))
display(data1)

features label Map(vectorType -> dense, length -> 1, values -> List(3504.0)) 18.0 Map(vectorType -> dense, length -> 1, values -> List(3693.0)) 15.0 Map(vectorType -> dense, length -> 1, values -> List(3436.0)) 18.0 Map(vectorType -> dense, length -> 1, values -> List(3433.0)) 16.0 Map(vectorType -> dense, length -> 1, values -> List(3449.0)) 17.0 Map(vectorType -> dense, length -> 1, values -> List(4341.0)) 15.0 Map(vectorType -> dense, length -> 1, values -> List(4354.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(4312.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(4425.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(3850.0)) 15.0 Map(vectorType -> dense, length -> 1, values -> List(3563.0)) 15.0 Map(vectorType -> dense, length -> 1, values -> List(3609.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(3761.0)) 15.0 Map(vectorType -> dense, length -> 1, values -> List(3086.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(2372.0)) 24.0 Map(vectorType -> dense, length -> 1, values -> List(2833.0)) 22.0 Map(vectorType -> dense, length -> 1, values -> List(2774.0)) 18.0 Map(vectorType -> dense, length -> 1, values -> List(2587.0)) 21.0 Map(vectorType -> dense, length -> 1, values -> List(2130.0)) 27.0 Map(vectorType -> dense, length -> 1, values -> List(1835.0)) 26.0 Map(vectorType -> dense, length -> 1, values -> List(2672.0)) 25.0 Map(vectorType -> dense, length -> 1, values -> List(2430.0)) 24.0 Map(vectorType -> dense, length -> 1, values -> List(2375.0)) 25.0 Map(vectorType -> dense, length -> 1, values -> List(2234.0)) 26.0 Map(vectorType -> dense, length -> 1, values -> List(2648.0)) 21.0 Map(vectorType -> dense, length -> 1, values -> List(4615.0)) 10.0 Map(vectorType -> dense, length -> 1, values -> List(4376.0)) 10.0 Map(vectorType -> dense, length -> 1, values -> List(4382.0)) 11.0 Map(vectorType -> dense, length -> 1, values -> List(4732.0)) 9.0 Map(vectorType -> dense, length -> 1, values -> List(2130.0)) 27.0 Map(vectorType -> dense, length -> 1, values -> List(2264.0)) 28.0 Map(vectorType -> dense, length -> 1, values -> List(2228.0)) 25.0 Map(vectorType -> dense, length -> 1, values -> List(2046.0)) 25.0 Map(vectorType -> dense, length -> 1, values -> List(2634.0)) 19.0 Map(vectorType -> dense, length -> 1, values -> List(3439.0)) 16.0 Map(vectorType -> dense, length -> 1, values -> List(3329.0)) 17.0 Map(vectorType -> dense, length -> 1, values -> List(3302.0)) 19.0 Map(vectorType -> dense, length -> 1, values -> List(3288.0)) 18.0 Map(vectorType -> dense, length -> 1, values -> List(4209.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(4464.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(4154.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(4096.0)) 14.0 Map(vectorType -> dense, length -> 1, values -> List(4955.0)) 12.0 Map(vectorType -> dense, length -> 1, values -> List(4746.0)) 13.0 Map(vectorType -> dense, length -> 1, values -> List(5140.0)) 13.0 Map(vectorType -> dense, length -> 1, values -> List(2962.0)) 18.0 Map(vectorType -> dense, length -> 1, values -> List(2408.0)) 22.0 Map(vectorType -> dense, length -> 1, values -> List(3282.0)) 19.0 Map(vectorType -> dense, length -> 1, values -> List(3139.0)) 18.0 Map(vectorType -> dense, length -> 1, values -> List(2220.0)) 23.0 Map(vectorType -> dense, length -> 1, values -> List(2123.0)) 28.0 Map(vectorType -> dense, length -> 1, values -> List(2074.0)) 30.0 Map(vectorType -> dense, length -> 1, values -> List(2065.0)) 30.0 Map(vectorType -> dense, length -> 1, values -> List(1773.0)) 31.0 Map(vectorType -> dense, length -> 1, values -> List(1613.0)) 35.0 Map(vectorType -> dense, length -> 1, values -> List(1834.0)) 27.0 Map(vectorType -> dense, length -> 1, values -> List(1955.0)) 26.0 Map(vectorType -> dense, length -> 1, values -> List(2278.0)) 24.0 Map(vectorType -> dense, length -> 1, values -> List(2126.0)) 25.0 Map(vectorType -> dense, length -

In [0]:
data2 = transData(dataMPG.select([dataMPG.columns[2], dataMPG.columns[4], dataMPG.columns[5],  dataMPG.columns[6], dataMPG.columns[0]]))
display(data2)

features label Map(vectorType -> dense, length -> 4, values -> List(307.0, 3504.0, 12.0, 70.0)) 18.0 Map(vectorType -> dense, length -> 4, values -> List(350.0, 3693.0, 11.5, 70.0)) 15.0 Map(vectorType -> dense, length -> 4, values -> List(318.0, 3436.0, 11.0, 70.0)) 18.0 Map(vectorType -> dense, length -> 4, values -> List(304.0, 3433.0, 12.0, 70.0)) 16.0 Map(vectorType -> dense, length -> 4, values -> List(302.0, 3449.0, 10.5, 70.0)) 17.0 Map(vectorType -> dense, length -> 4, values -> List(429.0, 4341.0, 10.0, 70.0)) 15.0 Map(vectorType -> dense, length -> 4, values -> List(454.0, 4354.0, 9.0, 70.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(440.0, 4312.0, 8.5, 70.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(455.0, 4425.0, 10.0, 70.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(390.0, 3850.0, 8.5, 70.0)) 15.0 Map(vectorType -> dense, length -> 4, values -> List(383.0, 3563.0, 10.0, 70.0)) 15.0 Map(vectorType -> dense, length -> 4, values -> List(340.0, 3609.0, 8.0, 70.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(400.0, 3761.0, 9.5, 70.0)) 15.0 Map(vectorType -> dense, length -> 4, values -> List(455.0, 3086.0, 10.0, 70.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(113.0, 2372.0, 15.0, 70.0)) 24.0 Map(vectorType -> dense, length -> 4, values -> List(198.0, 2833.0, 15.5, 70.0)) 22.0 Map(vectorType -> dense, length -> 4, values -> List(199.0, 2774.0, 15.5, 70.0)) 18.0 Map(vectorType -> dense, length -> 4, values -> List(200.0, 2587.0, 16.0, 70.0)) 21.0 Map(vectorType -> dense, length -> 4, values -> List(97.0, 2130.0, 14.5, 70.0)) 27.0 Map(vectorType -> dense, length -> 4, values -> List(97.0, 1835.0, 20.5, 70.0)) 26.0 Map(vectorType -> dense, length -> 4, values -> List(110.0, 2672.0, 17.5, 70.0)) 25.0 Map(vectorType -> dense, length -> 4, values -> List(107.0, 2430.0, 14.5, 70.0)) 24.0 Map(vectorType -> dense, length -> 4, values -> List(104.0, 2375.0, 17.5, 70.0)) 25.0 Map(vectorType -> dense, length -> 4, values -> List(121.0, 2234.0, 12.5, 70.0)) 26.0 Map(vectorType -> dense, length -> 4, values -> List(199.0, 2648.0, 15.0, 70.0)) 21.0 Map(vectorType -> dense, length -> 4, values -> List(360.0, 4615.0, 14.0, 70.0)) 10.0 Map(vectorType -> dense, length -> 4, values -> List(307.0, 4376.0, 15.0, 70.0)) 10.0 Map(vectorType -> dense, length -> 4, values -> List(318.0, 4382.0, 13.5, 70.0)) 11.0 Map(vectorType -> dense, length -> 4, values -> List(304.0, 4732.0, 18.5, 70.0)) 9.0 Map(vectorType -> dense, length -> 4, values -> List(97.0, 2130.0, 14.5, 71.0)) 27.0 Map(vectorType -> dense, length -> 4, values -> List(140.0, 2264.0, 15.5, 71.0)) 28.0 Map(vectorType -> dense, length -> 4, values -> List(113.0, 2228.0, 14.0, 71.0)) 25.0 Map(vectorType -> dense, length -> 4, values -> List(98.0, 2046.0, 19.0, 71.0)) 25.0 Map(vectorType -> dense, length -> 4, values -> List(232.0, 2634.0, 13.0, 71.0)) 19.0 Map(vectorType -> dense, length -> 4, values -> List(225.0, 3439.0, 15.5, 71.0)) 16.0 Map(vectorType -> dense, length -> 4, values -> List(250.0, 3329.0, 15.5, 71.0)) 17.0 Map(vectorType -> dense, length -> 4, values -> List(250.0, 3302.0, 15.5, 71.0)) 19.0 Map(vectorType -> dense, length -> 4, values -> List(232.0, 3288.0, 15.5, 71.0)) 18.0 Map(vectorType -> dense, length -> 4, values -> List(350.0, 4209.0, 12.0, 71.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(400.0, 4464.0, 11.5, 71.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(351.0, 4154.0, 13.5, 71.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(318.0, 4096.0, 13.0, 71.0)) 14.0 Map(vectorType -> dense, length -> 4, values -> List(383.0, 4955.0, 11.5, 71.0)) 12.0 Map(vectorType -> dense, length -> 4, values -> List(400.0, 4746.0, 12.0, 71.0)) 13.0 Map(vectorType -> dense, length -> 4, values -> List(400.0, 5140.0, 12.0, 71.0)) 13.0 Map(vectorType -> dense, length -> 4, values -> List(258.0, 2962.0, 13.5, 71.0)) 18.0 Map(vectorType -> dense, length -> 4, v

In [0]:
modelM1 = lr.fit(data1, {lr.regParam:0.0})
modelM2 = lr.fit(data2, {lr.regParam:0.0})

predictionsM1 = modelM1.transform(data1)
predictionsM2 = modelM2.transform(data2)

In [0]:
print(">>>> Model intercept: %r, coefficient: %r" % (modelM1.intercept, modelM1.coefficients))
print(">>>> Model intercept: %r, coefficient: %r" % (modelM2.intercept, modelM2.coefficients))

>>>> Model intercept: 46.31736442026568, coefficient: DenseVector([-0.0077])
>>>> Model intercept: -15.130597291474587, coefficient: DenseVector([0.0035, -0.007, 0.0791, 0.7553])


In [0]:
RMSE_M1 = evaluator.evaluate(predictionsM1)
RMSE_M2 = evaluator.evaluate(predictionsM2)

print("Model1: Root Mean Squared Error = " + str(RMSE_M1))
print("Model2: Root Mean Squared Error = " + str(RMSE_M2))

Model1: Root Mean Squared Error = 4.3336981591509565
Model2: Root Mean Squared Error = 3.417052956459645


### Gradient Boosted Tree Regression

Lets try another model from the spark toolkit on our data: Gradient Boosted Tree Regression https://statisticasoftware.wordpress.com/2012/09/11/boosting-trees-for-regression-and-classification/

Follow the steps below to see how easily we can apply a different machine learning approach to our data.

In [0]:
# Import LinearRegression class
from pyspark.ml.regression import GBTRegressor

# Define LinearRegression algorithm
rf = GBTRegressor(maxDepth=2, seed=42) #1st try: maxDepth=2, seed=42 | 2nd try: maxDepth=2, seed=42 
dataMPGOrdered = transData(dataMPG.select(dataMPG.columns[1:] + [dataMPG.columns[0]]  ))
rf.setMaxIter(10) #1st try: 10 | 2nd try: 30
rf.setMinWeightFractionPerNode(0.1)  #1st try: 0.1 | 2nd try: 0.005
modelG = rf.fit(dataMPGOrdered)

In [0]:
predictions = modelG.transform(dataMPGOrdered)
predictions.select("features","label", "prediction").show(5)

+--------------------+-----+------------------+
|            features|label|        prediction|
+--------------------+-----+------------------+
|[8.0,307.0,130.0,...| 18.0|13.850634402202939|
|[8.0,350.0,165.0,...| 15.0|13.850634402202939|
|[8.0,318.0,150.0,...| 18.0|13.850634402202939|
|[8.0,304.0,150.0,...| 16.0|13.850634402202939|
|[8.0,302.0,140.0,...| 17.0|13.850634402202939|
+--------------------+-----+------------------+
only showing top 5 rows



In [0]:
# Select (prediction, true label) and compute test error
evaluator = RegressionEvaluator(
    labelCol="label", predictionCol="prediction", metricName="rmse")
rmse = evaluator.evaluate(predictions)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmse)

#1st 3.12151  | 2nd 0.847873

Root Mean Squared Error (RMSE) on test data = 3.15872


How did this approach perform in comparison to the previous regressions? Remember we are only looking at RMSE on the training data.

This model is prone for overfitting, below we split the data into a training and test dataset. This would be standard good practice if this is all of the data we have.

Fit the model to the training dataset and test the fit against the test dataset.

In [0]:
(trainingData, testData) = dataMPGOrdered.randomSplit([0.6, 0.4])

trainingData.show(5)
testData.show(5)

+--------------------+-----+
|            features|label|
+--------------------+-----+
|[3.0,70.0,90.0,21...| 18.0|
|[3.0,70.0,97.0,23...| 19.0|
|[3.0,70.0,100.0,2...| 23.7|
|[3.0,80.0,110.0,2...| 21.5|
|[4.0,68.0,49.0,18...| 29.0|
+--------------------+-----+
only showing top 5 rows

+--------------------+-----+
|            features|label|
+--------------------+-----+
|[4.0,72.0,69.0,16...| 35.0|
|[4.0,78.0,52.0,19...| 32.8|
|[4.0,79.0,67.0,19...| 26.0|
|[4.0,83.0,61.0,20...| 32.0|
|[4.0,85.0,52.0,20...| 29.0|
+--------------------+-----+
only showing top 5 rows



In [0]:
modelGTr = rf.fit(trainingData)
predictionsGTr = modelGTr.transform(testData)
rmseGtr = evaluator.evaluate(predictionsGTr)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmseGtr)

Root Mean Squared Error (RMSE) on test data = 3.92357


## Bonus exercise: Machine learning practice

I've included below a dataset of wine characteristics and quality, you can use this to practice applying the aprroachs above to build a predictor of wine quality.

Try to find the best model you can to predict wine quality.

If you want, you can try other machine learning models, see: https://spark.apache.org/docs/2.2.0/ml-classification-regression.html#regression

##### Below I fit two basic models to the data, based on the methods we have seen in this lab, and compare them. However, there is a lot more you could do.

In [0]:
%sh wget https://raw.githubusercontent.com/MingChen0919/learning-apache-spark/master/data/WineData.csv

--2024-05-18 19:30:34--  https://raw.githubusercontent.com/MingChen0919/learning-apache-spark/master/data/WineData.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 84143 (82K) [text/plain]
Saving to: ‘WineData.csv’

     0K .......... .......... .......... .......... .......... 60% 6.39M 0s
    50K .......... .......... .......... ..                   100% 5.59M=0.01s

2024-05-18 19:30:34 (6.05 MB/s) - ‘WineData.csv’ saved [84143/84143]



In [0]:
wineDF = spark.read.format("csv").option("header", "true").option("inferSchema", "true").load("file:/databricks/driver/WineData.csv")
wineDF.cache()
wineDF.printSchema()
wineDF.select("quality").distinct().take(100)

root
 |-- fixed acidity: double (nullable = true)
 |-- volatile acidity: double (nullable = true)
 |-- citric acid: double (nullable = true)
 |-- residual sugar: double (nullable = true)
 |-- chlorides: double (nullable = true)
 |-- free sulfur dioxide: double (nullable = true)
 |-- total sulfur dioxide: double (nullable = true)
 |-- density: double (nullable = true)
 |-- pH: double (nullable = true)
 |-- sulphates: double (nullable = true)
 |-- alcohol: double (nullable = true)
 |-- quality: integer (nullable = true)

Out[30]: [Row(quality=6),
 Row(quality=3),
 Row(quality=5),
 Row(quality=4),
 Row(quality=8),
 Row(quality=7)]

In [0]:
# Convert to float format
def string_to_float(x):
    return float(x)

#
def condition(r):
    if (0<= r <= 4):
        label = "low"
    elif(4< r <= 6):
        label = "medium"
    else:
        label = "high"
    return label
  
from pyspark.sql.functions import udf
from pyspark.sql.types import StringType, DoubleType
string_to_float_udf = udf(string_to_float, DoubleType())
quality_udf = udf(lambda x: condition(x), StringType())
wineDFQ = wineDF.withColumn("quality", quality_udf("quality"))

In [0]:
wineDF.show(10,True)
wineDF.printSchema()

+-------------+----------------+-----------+--------------+---------+-------------------+--------------------+-------+----+---------+-------+-------+
|fixed acidity|volatile acidity|citric acid|residual sugar|chlorides|free sulfur dioxide|total sulfur dioxide|density|  pH|sulphates|alcohol|quality|
+-------------+----------------+-----------+--------------+---------+-------------------+--------------------+-------+----+---------+-------+-------+
|          7.4|             0.7|        0.0|           1.9|    0.076|               11.0|                34.0| 0.9978|3.51|     0.56|    9.4|      5|
|          7.8|            0.88|        0.0|           2.6|    0.098|               25.0|                67.0| 0.9968| 3.2|     0.68|    9.8|      5|
|          7.8|            0.76|       0.04|           2.3|    0.092|               15.0|                54.0|  0.997|3.26|     0.65|    9.8|      5|
|         11.2|            0.28|       0.56|           1.9|    0.075|               17.0|           

In [0]:
dataWine = transData(wineDF)
display(dataWine)

features label Map(vectorType -> dense, length -> 11, values -> List(7.4, 0.7, 0.0, 1.9, 0.076, 11.0, 34.0, 0.9978, 3.51, 0.56, 9.4)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.8, 0.88, 0.0, 2.6, 0.098, 25.0, 67.0, 0.9968, 3.2, 0.68, 9.8)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.8, 0.76, 0.04, 2.3, 0.092, 15.0, 54.0, 0.997, 3.26, 0.65, 9.8)) 5 Map(vectorType -> dense, length -> 11, values -> List(11.2, 0.28, 0.56, 1.9, 0.075, 17.0, 60.0, 0.998, 3.16, 0.58, 9.8)) 6 Map(vectorType -> dense, length -> 11, values -> List(7.4, 0.7, 0.0, 1.9, 0.076, 11.0, 34.0, 0.9978, 3.51, 0.56, 9.4)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.4, 0.66, 0.0, 1.8, 0.075, 13.0, 40.0, 0.9978, 3.51, 0.56, 9.4)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.9, 0.6, 0.06, 1.6, 0.069, 15.0, 59.0, 0.9964, 3.3, 0.46, 9.4)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.3, 0.65, 0.0, 1.2, 0.065, 15.0, 21.0, 0.9946, 3.39, 0.47, 10.0)) 7 Map(vectorType -> dense, length -> 11, values -> List(7.8, 0.58, 0.02, 2.0, 0.073, 9.0, 18.0, 0.9968, 3.36, 0.57, 9.5)) 7 Map(vectorType -> dense, length -> 11, values -> List(7.5, 0.5, 0.36, 6.1, 0.071, 17.0, 102.0, 0.9978, 3.35, 0.8, 10.5)) 5 Map(vectorType -> dense, length -> 11, values -> List(6.7, 0.58, 0.08, 1.8, 0.097, 15.0, 65.0, 0.9959, 3.28, 0.54, 9.2)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.5, 0.5, 0.36, 6.1, 0.071, 17.0, 102.0, 0.9978, 3.35, 0.8, 10.5)) 5 Map(vectorType -> dense, length -> 11, values -> List(5.6, 0.615, 0.0, 1.6, 0.089, 16.0, 59.0, 0.9943, 3.58, 0.52, 9.9)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.8, 0.61, 0.29, 1.6, 0.114, 9.0, 29.0, 0.9974, 3.26, 1.56, 9.1)) 5 Map(vectorType -> dense, length -> 11, values -> List(8.9, 0.62, 0.18, 3.8, 0.176, 52.0, 145.0, 0.9986, 3.16, 0.88, 9.2)) 5 Map(vectorType -> dense, length -> 11, values -> List(8.9, 0.62, 0.19, 3.9, 0.17, 51.0, 148.0, 0.9986, 3.17, 0.93, 9.2)) 5 Map(vectorType -> dense, length -> 11, values -> List(8.5, 0.28, 0.56, 1.8, 0.092, 35.0, 103.0, 0.9969, 3.3, 0.75, 10.5)) 7 Map(vectorType -> dense, length -> 11, values -> List(8.1, 0.56, 0.28, 1.7, 0.368, 16.0, 56.0, 0.9968, 3.11, 1.28, 9.3)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.4, 0.59, 0.08, 4.4, 0.086, 6.0, 29.0, 0.9974, 3.38, 0.5, 9.0)) 4 Map(vectorType -> dense, length -> 11, values -> List(7.9, 0.32, 0.51, 1.8, 0.341, 17.0, 56.0, 0.9969, 3.04, 1.08, 9.2)) 6 Map(vectorType -> dense, length -> 11, values -> List(8.9, 0.22, 0.48, 1.8, 0.077, 29.0, 60.0, 0.9968, 3.39, 0.53, 9.4)) 6 Map(vectorType -> dense, length -> 11, values -> List(7.6, 0.39, 0.31, 2.3, 0.082, 23.0, 71.0, 0.9982, 3.52, 0.65, 9.7)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.9, 0.43, 0.21, 1.6, 0.106, 10.0, 37.0, 0.9966, 3.17, 0.91, 9.5)) 5 Map(vectorType -> dense, length -> 11, values -> List(8.5, 0.49, 0.11, 2.3, 0.084, 9.0, 67.0, 0.9968, 3.17, 0.53, 9.4)) 5 Map(vectorType -> dense, length -> 11, values -> List(6.9, 0.4, 0.14, 2.4, 0.085, 21.0, 40.0, 0.9968, 3.43, 0.63, 9.7)) 6 Map(vectorType -> dense, length -> 11, values -> List(6.3, 0.39, 0.16, 1.4, 0.08, 11.0, 23.0, 0.9955, 3.34, 0.56, 9.3)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.6, 0.41, 0.24, 1.8, 0.08, 4.0, 11.0, 0.9962, 3.28, 0.59, 9.5)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.9, 0.43, 0.21, 1.6, 0.106, 10.0, 37.0, 0.9966, 3.17, 0.91, 9.5)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.1, 0.71, 0.0, 1.9, 0.08, 14.0, 35.0, 0.9972, 3.47, 0.55, 9.4)) 5 Map(vectorType -> dense, length -> 11, values -> List(7.8, 0.645, 0.0, 2.0, 0.082, 8.0, 16.0, 0.9964, 3.38, 0.59, 9.8)) 6 Map(vectorType -> dense, length -> 11, values -> List(6.7, 0.675, 0.07, 2.4, 0.089, 17.0, 82.0, 0.9958, 3.35, 0.54, 10.1)) 5 Map(vectorType -> dense, length -> 11, values -> List(6.9, 0.685, 0.0, 2.5, 0.105, 22.0, 37.0, 0.9966, 3.46, 0.57, 10.6)) 6 Map(vectorType -> dense, length -> 11, values -> List(8.3, 0.655, 

In [0]:
(trainingDataWine, testDataWine) = dataWine.randomSplit([0.6, 0.4])

In [0]:
# Gradient Boosted Trees
modelGBTWine = rf.fit(trainingDataWine)

In [0]:
predictionsGBTWine = modelGBTWine.transform(testDataWine)

In [0]:
rmseGBTWine = evaluator.evaluate(predictionsGBTWine)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmseGBTWine)

Root Mean Squared Error (RMSE) on test data = 0.663375


In [0]:
# Fit 2 models, using different regularization parameters
modelWineA = lr.fit(trainingDataWine, {lr.regParam:0.0})
modelWineB = lr.fit(trainingDataWine, {lr.regParam:100.0})


In [0]:
predictionsWineA = modelWineA.transform(testDataWine)
predictionsWineB = modelWineB.transform(testDataWine)

In [0]:
RMSE_WineA = evaluator.evaluate(predictionsWineA)
RMSE_WineB = evaluator.evaluate(predictionsWineB)

print("ModelA: Root Mean Squared Error = " + str(RMSE_WineA))
print("ModelB: Root Mean Squared Error = " + str(RMSE_WineB))

ModelA: Root Mean Squared Error = 0.6468707308612757
ModelB: Root Mean Squared Error = 0.8106450132073024
